src/hierarchies/base_hierarchy.h

```diff
@@ -29,101 +29,312 @@
-  virtual double marg_lpdf(
-      const bool posterior, const Eigen::RowVectorXd &datum,
-      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) const = 0;
+  virtual double prior_pred_lpdf(
+      const Eigen::RowVectorXd &datum,
+      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) = 0;
+
+  virtual double conditional_pred_lpdf(
+      const Eigen::RowVectorXd &datum,
+      const Eigen::RowVectorXd &covariate = Eigen::VectorXd(0)) = 0;
+
   // EVALUATION FUNCTIONS FOR GRIDS OF POINTS
   //! Evaluates the log-likelihood of data in a grid of points
   virtual Eigen::VectorXd like_lpdf_grid(
       const Eigen::MatrixXd &data,
-      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) const;
+      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) = 0;
   //! Evaluates the log-marginal of data in a grid of points
-  virtual Eigen::VectorXd marg_lpdf_grid(
-      const bool posterior, const Eigen::MatrixXd &data,
-      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) const;
+  virtual Eigen::VectorXd prior_pred_lpdf_grid(
+      const Eigen::MatrixXd &data,
+      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) = 0;
+
+  virtual Eigen::VectorXd conditional_pred_lpdf_grid(
+      const Eigen::MatrixXd &data,
+      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) = 0;
 
   // SAMPLING FUNCTIONS
   //! Generates new values for state from the centering prior distribution
-  virtual void draw() = 0;
+  virtual void sample_prior() = 0;
   //! Generates new values for state from the centering posterior distribution
-  virtual void sample_given_data(bool update_params = true) = 0;
-  //! Overloaded version of sample_given_data(), mainly used for debugging
-  void sample_given_data(
-      const Eigen::MatrixXd &data,
-      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0));
-
-  // GETTERS AND SETTERS
-  google::protobuf::Message *get_mutable_prior() {
-    if (prior == nullptr) create_empty_prior();
-
-    return prior.get();
-  }
+  virtual void sample_full_cond(bool update_params = false) = 0;
   virtual void write_state_to_proto(google::protobuf::Message *out) const = 0;
   virtual void write_hypers_to_proto(google::protobuf::Message *out) const = 0;
   virtual void set_state_from_proto(
       const google::protobuf::Message &state_) = 0;
+
+  virtual void check_prior_is_set() = 0;
+
+  // GETTERS AND SETTERS
+  int get_card() const { return card; }
+  double get_log_card() const { return log_card; }
+  std::set<int> get_data_idx() { return cluster_data_idx; }
+  virtual google::protobuf::Message *get_mutable_prior() = 0;
+
+  //! Overloaded version of sample_full_cond(), mainly used for debugging
+  virtual void sample_full_cond(
+      const Eigen::MatrixXd &data,
+      const Eigen::MatrixXd &covariates = Eigen::MatrixXd(0, 0)) = 0;
+};
+
+template <class Derived, typename State, typename Hyperparams, typename Prior>
+class BaseHierarchy : public AbstractHierarchy {
```

Comment:
Is it OK to declare these two classes in the same file? It feels inappropriate to me, but again, I don't know anything